Add the support for storing a shared process for site isolation in process cache

rniwa opened the pull request:
#### 8cef11213b35473894605d119a6a213127eb0c8d
<pre>
Add the support for storing a shared process for site isolation in process cache
<a href="https://bugs.webkit.org/show_bug.cgi?id=299916">https://bugs.webkit.org/show_bug.cgi?id=299916</a>

Reviewed by Ben Nham.

Made shared process for site isolation work with the Web process cache. WebProcessCache is extended to
maintain a separate map of shared processes per top-level domain in addition to the existing cache of
&quot;main&quot; processes. The size and the capacity of WebProcessCache now takes into account both types of the
process caches. In addition to being segregated by the top-level domain, when a given domain receives
an user interaction from the user, the existing shared process cache which had loaded the same domain
will no longer be eligible for the cache reuse.

This PR also fixes the bug that a wrong BrowsingContextGroup is used during a back/forward navigation.
Store the relevant BrowsingContextGroup in WebBackForwardListItem and use this BrowsingContextGroup
when back/forward navigating to the item.

In addition, this PR also fixes the bug in WebPageProxy::receivedNavigationActionPolicyDecision that
we were creating a new BrowsingContextGroup even when the result of redirect chain ended up navigating
back to the same site. We now avoid creating a new BrowsingContextGroup when main frame&apos;s site did not
change per navigation.

Test: TestWebKitAPI.SiteIsolation.SharedProcessBasicWebProcessCache
      TestWebKitAPI.SiteIsolation.SharedProcessWebProcessCacheSharedProcessForSiteWithUserInteraction

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm
* Source/WebKit/Shared/WebBackForwardListItem.cpp:
(WebKit::WebBackForwardListItem::create): Optionally store BrowsingContextGroup.
(WebKit::WebBackForwardListItem::WebBackForwardListItem):
* Source/WebKit/Shared/WebBackForwardListItem.h:
(WebKit::WebBackForwardListItem::browsingContextGroup const):
* Source/WebKit/UIProcess/BrowsingContextGroup.cpp:
(WebKit::BrowsingContextGroup::sharedProcessForSite):
(WebKit::BrowsingContextGroup::ensureProcessForSite):
* Source/WebKit/UIProcess/BrowsingContextGroup.h:
* Source/WebKit/UIProcess/FrameProcess.cpp:
(WebKit::FrameProcess::FrameProcess):
* Source/WebKit/UIProcess/FrameProcess.h:
(WebKit::FrameProcess::sharedProcessMainFrameSite const):
(WebKit::FrameProcess::create):
* Source/WebKit/UIProcess/WebBackForwardList.cpp:
(WebKit::WebBackForwardList::backForwardAddItemShared):
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::prepareForProvisionalLoadInProcess):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::launchProcess):
(WebKit::WebPageProxy::initializeWebPage):
(WebKit::WebPageProxy::receivedNavigationActionPolicyDecision): Fixed a subtle bug that this code was
using a wrong BrowsingContextGroup to find the shared process when the navigation is supposed to swap
the BrowsingContextGroup.
(WebKit::WebPageProxy::continueNavigationInNewProcess):
(WebKit::WebPageProxy::triggerBrowsingContextGroupSwitchForNavigation):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebProcessCache.cpp: Improved the logging code in this file to use
SENSITIVE_LOG_STRING of PRIVATE_LOG_STRING since it logs the site name.
(WebKit::WebProcessCache::canCacheProcess const):
(WebKit::WebProcessCache::addProcess):
(WebKit::WebProcessCache::evictAtRandomIfNeeded): Extracted from addProcess. We try to evict a shared
process before trying to evict the &quot;main&quot; process since the main process needs to launch sooner.
(WebKit::WebProcessCache::takeProcess):
(WebKit::WebProcessCache::takeSharedProcess): Added.
* Source/WebKit/UIProcess/WebProcessCache.h:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::processForSite): Added the logic to reuse a cached shared process.
(WebKit::WebProcessPool::createWebPage):
(WebKit::WebProcessPool::processForNavigation):
(WebKit::WebProcessPool::prepareProcessForNavigation):
(WebKit::WebProcessPool::processForNavigationInternal):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::addSharedProcessDomain): Added.
(WebKit::WebProcessProxy::didStartUsingProcessForSiteIsolation):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::siteIsolatedViewWithSharedProcess):
(TestWebKitAPI::(SiteIsolation, SharedProcessBasicWebProcessCache)): Added
(TestWebKitAPI::(SiteIsolation, SharedProcessWithResourceLoadStatistics)):
(TestWebKitAPI::(SiteIsolation, SharedProcessAfterClick)):
(TestWebKitAPI::(SiteIsolation, SharedProcessAfterKeyDown)):
(TestWebKitAPI::(SiteIsolation, SharedProcessAfterUserInteractionInSharedProcesss)):
(TestWebKitAPI::(SiteIsolation, SharedProcessWebProcessCacheSharedProcessForSiteWithUserInteraction)):
Added.

Canonical link: <a href="https://commits.webkit.org/300964@main">https://commits.webkit.org/300964@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b4150a2e0ef122eb5f4fbd034eb3e92c51c5016a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124435 "10 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44119 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34850 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131279 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/76450 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/f9cbaa04-883a-481f-86bb-18752120b282) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/126312 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44862 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52716 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94668 "Failed to checkout and rebase branch from PR 51603") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/76450 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127389 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35763 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111299 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75244 "Passed tests") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/0072647b-d3fa-4740-8e2a-df8879ef1711) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34699 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29459 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74758 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105515 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29679 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133944 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51335 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39163 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103145 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51736 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107514 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102934 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48299 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26555 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/48281 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19538 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51201 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56988 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50619 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53981 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52295 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/UIProcess/BrowsingContextGroup.cpp:
```diff
@@ -45,7 +45,7 @@ BrowsingContextGroup::BrowsingContextGroup() = default;
 BrowsingContextGroup::~BrowsingContextGroup() = default;
 
-void BrowsingContextGroup::sharedProcessForSite(WebsiteDataStore& websiteDataStore, API::WebsitePolicies* websitePolicies, const WebPreferences& preferences, const WebCore::Site& site,
+void BrowsingContextGroup::sharedProcessForSite(WebsiteDataStore& websiteDataStore, API::WebsitePolicies* websitePolicies, const WebPreferences& preferences, const WebCore::Site& site, const WebCore::Site& mainFrameSite,
```
Maybe we should make something like ClientOrigin at some point but with Sites instead of SecurityOriginData

rniwa replied:
Maybe?

achristensen07 commented on Source/WebKit/UIProcess/BrowsingContextGroup.cpp:
```diff
@@ -58,7 +58,7 @@ void BrowsingContextGroup::sharedProcessForSite(WebsiteDataStore& websiteDataSto
             return completionHandler(nullptr);
     }
-    websiteDataStore.fetchDomainsWithUserInteraction([protectedThis = Ref { *this }, websiteDataStore = Ref { websiteDataStore }, preferences = Ref { preferences }, site = Site { site },
+    websiteDataStore.fetchDomainsWithUserInteraction([protectedThis = Ref { *this }, websiteDataStore = Ref { websiteDataStore }, preferences = Ref { preferences }, site = Site { site }, mainFrameSite = Site { mainFrameSite },
```
I've started to take long capture lists like this and put them each on their own line, like continueWithProcessForNavigation

rniwa replied:
Sure, we can do that.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -5128,8 +5128,9 @@ void WebPageProxy::receivedNavigationActionPolicyDecision(WebProcessProxy& proce
-        [this, protectedThis = Ref { *this }, frame = Ref { frame }, navigation = Ref { navigation }, site, sourceURL, processSwapRequestedByClient, lockdownMode, enhancedSecurity, loadedWebArchive, frameInfo = FrameInfoData { frameInfo }, websiteDataStore = websiteDataStore.copyRef(), continueWithProcessForNavigation = WTFMove(continueWithProcessForNavigation)](FrameProcess* sharedProcess) mutab ...
+    Site mainFrameSite = frame.isMainFrame() ? site : Site { URL(protectedPageLoadState()->activeURL()) };
+    browsingContextGroup->sharedProcessForSite(websiteDataStore, policies.get(), preferences, site, mainFrameSite, lockdownMode, enhancedSecurity, Ref { m_configuration }, frame.isMainFrame() ? IsMainFrame::Yes : IsMainFrame::No,
+        [this, protectedThis = Ref { *this }, frame = Ref { frame }, navigation = Ref { navigation }, site, mainFrameSite, sourceURL, processSwapRequestedByClient, lockdownMode, enhancedSecurity, loadedWebArchive, frameInfo = FrameInfoData { frameInfo }, websiteDataStore = websiteDataStore.copyRef(), continueWithProcessForNavigation = WTFMove(continueWithProcessForNavigation)](FrameProcess* share ...
```
Ditto.  Too many captures.

rniwa replied:
Fixed.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -8600,8 +8601,8 @@ void WebPageProxy::triggerBrowsingContextGroupSwitchForNavigation(WebCore::Navig
         processForNavigation = m_configuration->protectedProcessPool()->createNewWebProcess(protectedWebsiteDataStore().ptr(), lockdownMode, enhancedSecurity, WebProcessProxy::IsPrewarmed::No, CrossOriginMode::Isolated);
-    else
-        processForNavigation = m_configuration->protectedProcessPool()->processForSite(protectedWebsiteDataStore(), responseSite, lockdownMode, enhancedSecurity, m_configuration, WebCore::ProcessSwapDisposition::COOP);
+    else // FIXME: Is this for main frame only?
```
Yes, it should be.

rniwa replied:
Oh, I've already confirmed this. Will remove FIXME.

achristensen07 commented on Source/WebKit/UIProcess/WebProcessProxy.h:
```diff
@@ -769,6 +772,7 @@ class WebProcessProxy final : public AuxiliaryProcessProxy {
     HashMap<String, uint64_t> m_pageURLRetainCountMap;
 
     Expected<WebCore::Site, SiteState> m_site { Unexpected<SiteState> { SiteState::NotYetSpecified } };
+    std::optional<WebCore::Site> m_sharedProcessSite;
```
m_sharedProcessMainFrameSite?

rniwa replied:
Sure, will rename.

bnham commented on Source/WebKit/UIProcess/BrowsingContextGroup.cpp:
```diff
@@ -68,30 +68,30 @@ void BrowsingContextGroup::sharedProcessForSite(WebsiteDataStore& websiteDataSto
-        Ref process = pageConfiguration->protectedProcessPool()->processForSite(websiteDataStore.get(), site, lockdownMode, enhancedSecurity, pageConfiguration.get(), ProcessSwapDisposition::Other);
-        Ref frameProcess = FrameProcess::create(process, protectedThis, std::nullopt, preferences, InjectBrowsingContextIntoProcess::Yes);
+        Ref process = pageConfiguration->protectedProcessPool()->processForSite(websiteDataStore.get(), WebProcessPool::IsSharedProcess::Yes, site, mainFrameSite, lockdownMode, enhancedSecurity, pageConfiguration.get(), ProcessSwapDisposition::Other);
+        Ref frameProcess = FrameProcess::create(process, protectedThis, std::nullopt, mainFrameSite, preferences, InjectBrowsingContextIntoProcess::Yes);
```
I think there's some missing logic here. Suppose `processForSite` returns a cached shared process here. We need to intersect the set of sites already loaded in to the cached shared process with the set of sites that have a user interaction. If the set is non-empty then we have to discard the cached shared process and load the site in to a new shared process.

Put another way:

* You start interacting with foo.com as a first party, therefore we want foo.com to be isolated in any future load.
* You load the bar.com page which has a foo.com iframe.
  * We find a cached shared process for bar.com.
  * Let's say this cached shared process already has foo.com loaded in to it.

At this point, I think this patch will try to use the cached shared process it found. But it actually has to discard the cached shared process and isolate foo.com in to its own process.

rniwa replied:
Good point. We should add a test for that too.

bnham commented on Source/WebKit/UIProcess/BrowsingContextGroup.cpp:
```diff
@@ -68,30 +68,30 @@ void BrowsingContextGroup::sharedProcessForSite(WebsiteDataStore& websiteDataSto
 }
 
-FrameProcess* BrowsingContextGroup::processForSite(const Site& site)
+FrameProcess* BrowsingContextGroup::processForSite(const Site& site, const Site& /* mainFrameSite */)
```
Seems like mainFrameSite does not need to be a parameter here.

rniwa replied:
oh yeah, will remove.

bnham commented on Source/WebKit/UIProcess/WebProcessCache.cpp:
```diff
@@ -154,30 +154,53 @@ bool WebProcessCache::addProcess(Ref<CachedProcess>&& cachedProcess)
+
+        evictAtRandomIfNeeded();
+
+        WEBPROCESSCACHE_RELEASE_LOG("addProcess: Added shared process to WebProcess cache (size=%u, capacity=%u) %" PRIVATE_LOG_STRING, cachedProcess->process().processID(), size() + 1, capacity(), site->toString().utf8().data());
```
Since it's a RELEASE_LOG for a URL-like string, this probably needs to be SENSITIVE_LOG_STRING.

rniwa replied:
Sure, I guess I'll also fix the existing code that's using PRIVATE_LOG_STRING.

bnham commented on Source/WebKit/UIProcess/WebProcessCache.cpp:
```diff
@@ -221,6 +244,49 @@ RefPtr<WebProcessProxy> WebProcessCache::takeProcess(const WebCore::Site& site,
+    }
+
+    Ref process = m_sharedProcessesPerSite.take(it)->takeProcess();
+    WEBPROCESSCACHE_RELEASE_LOG("takeSharedProcess: Taking process from WebProcess cache (size=%u, capacity=%u, processWasTerminated=%d) %" PRIVATE_LOG_STRING, process->processID(), size(), capacity(), process->wasTerminated(), mainFrameSite.toString().utf8().data());
```
`SENSITIVE_LOG_STRING`

rniwa replied:
Fixed.

bnham commented on Source/WebKit/UIProcess/WebProcessCache.cpp:
```diff
@@ -154,30 +154,53 @@ bool WebProcessCache::addProcess(Ref<CachedProcess>&& cachedProcess)
 
+void WebProcessCache::evictAtRandomIfNeeded()
+{
+    while (size() >= capacity()) {
```
If foo.com is the main site it seems like we could end up in a situation where foo.com's main site process is not in the process cache, but foo.com's shared process is in the shared process cache (based on other logic in WebKit that determines whether a process can be added to the process cache or not). In that case we end up with shared processes in the cache that basically can't be evicted.

Maybe we need to do a weighted random selection of evicting from either m_sharedProcessesPerSite or m_processesPerSite instead of always evicting based on m_processesPerSite.

rniwa replied:
Hm... I think we want to prefer evicting the shared process first though. I guess I'm gonna randomly pick from `m_sharedProcessesPerSite` then randomly pick from `m_processesPerSite` and take turns thereafter.

bnham replied:
That also works, we just need to make sure to also consult `m_sharedProcessesPerSite` in this function.

achristensen07 reviewed (changes requested):
Looks like some tests need to be fixed.  Removing approval...

bnham commented on Source/WebKit/UIProcess/BrowsingContextGroup.cpp:
```diff
@@ -68,17 +77,20 @@ void BrowsingContextGroup::sharedProcessForSite(WebsiteDataStore& websiteDataSto
 }
 
-Ref<FrameProcess> BrowsingContextGroup::ensureProcessForSite(const Site& site, WebProcessProxy& process, const WebPreferences& preferences, InjectBrowsingContextIntoProcess injectBrowsingContextIntoProcess)
+Ref<FrameProcess> BrowsingContextGroup::ensureProcessForSite(const Site& site, const Site& mainFrameSite, WebProcessProxy& process, const WebPreferences& preferences, InjectBrowsingContextIntoProcess injectBrowsingContextIntoProcess)
```
When are you supposed to use `ensureProcessForSite` instead of `sharedProcessForSite`? I tried reading through the callsites and don't understand the logic.

rniwa replied:
So `sharedProcessForSite` is meant to be called in exactly one place, `WebPageProxy::receivedNavigationActionPolicyDecision`, to decide whether to use a shared process or not. Once we've determined that, we should be using `ensureProcessForSite` instead.

bnham commented on Source/WebKit/UIProcess/WebProcessCache.cpp:
```diff
@@ -154,35 +154,65 @@ bool WebProcessCache::addProcess(Ref<CachedProcess>&& cachedProcess)
+                break;
+        }
+        auto it = m_processesPerSite.random();
+        if (RefPtr sharedProcess = m_sharedProcessesPerSite.take(it->key)) {
+            WEBPROCESSCACHE_RELEASE_LOG("addProcess: Evicting shared process from WebProcess cache because capacity was reached", it->value->process().processID());
+            if (size() < capacity())
+                break;
+        }
```
Passing iterator in to wrong hash map. It looks like this entire block is a leftover copypasta that should be deleted.

rniwa replied:
Oh, no. We're using `it->key` which is the `WebCore::Site`. The idea here is to always evict the shared process first before we evict the main process for any given `WebCore::Site`.

bnham replied:
I see, makes sense.

bnham commented on Source/WebKit/UIProcess/WebProcessCache.cpp:
```diff
@@ -207,7 +237,55 @@ RefPtr<WebProcessProxy> WebProcessCache::takeProcess(const WebCore::Site& site,
+    if (!it->value->process().sharedProcessDomains().intersectionWith(isolatedDomains).isEmpty()) {
+        WEBPROCESSCACHE_RELEASE_LOG("takeSharedProcess: cannot take process, contains a domain that needs to be isolated", it->value->process().processID());
+        return nullptr;
+    }
```
It seems like at this point, instead of just returning `nullptr`, you would also want to remove the process from the cache, since it seems like there's no way we could ever use that cached process in the future.

It looks like you could probably also move this site check out of WebProcessCache and up to one of the callers of this method. If you did that, then you could just drop the process that you've already taken out of the process cache on the floor at whatever point you do the site check.

For instance it seems like you could move the site check all the way up to `BrowsingContextGroup::sharedProcessForSite`. Then you wouldn't have to propagate the sites-with-interaction list all the way down to WebProcessCache. Then, if `BrowsingContextGroup::sharedProcessForSite` discovers that the cached process fails the site intersection check, it can drop the cached process on the floor at that point.

rniwa replied:
Ok, I'll move up to `WebProcessPool::processForSite` since that's where we create a new process if there is no match in the cache.

bnham reviewed (changes requested):
Looks mostly there, definitely need to fix the issue in cache eviction though.

rniwa committed:
Add the support for storing a shared process for site isolation in process cache
https://bugs.webkit.org/show_bug.cgi?id=299916

Reviewed by Ben Nham.

Made shared process for site isolation work with the Web process cache. WebProcessCache is extended to
maintain a separate map of shared processes per top-level domain in addition to the existing cache of
"main" processes. The size and the capacity of WebProcessCache now takes into account both types of the
process caches. In addition to being segregated by the top-level domain, when a given domain receives
an user interaction from the user, the existing shared process cache which had loaded the same domain
will no longer be eligible for the cache reuse.

This PR also fixes the bug that a wrong BrowsingContextGroup is used during a back/forward navigation.
Store the relevant BrowsingContextGroup in WebBackForwardListItem and use this BrowsingContextGroup
when back/forward navigating to the item.

In addition, this PR also fixes the bug in WebPageProxy::receivedNavigationActionPolicyDecision that
we were creating a new BrowsingContextGroup even when the result of redirect chain ended up navigating
back to the same site. We now avoid creating a new BrowsingContextGroup when main frame's site did not
change per navigation.

Test: TestWebKitAPI.SiteIsolation.SharedProcessBasicWebProcessCache
      TestWebKitAPI.SiteIsolation.SharedProcessWebProcessCacheSharedProcessForSiteWithUserInteraction

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm
* Source/WebKit/Shared/WebBackForwardListItem.cpp:
(WebKit::WebBackForwardListItem::create): Optionally store BrowsingContextGroup.
(WebKit::WebBackForwardListItem::WebBackForwardListItem):
* Source/WebKit/Shared/WebBackForwardListItem.h:
(WebKit::WebBackForwardListItem::browsingContextGroup const):
* Source/WebKit/UIProcess/BrowsingContextGroup.cpp:
(WebKit::BrowsingContextGroup::sharedProcessForSite):
(WebKit::BrowsingContextGroup::ensureProcessForSite):
* Source/WebKit/UIProcess/BrowsingContextGroup.h:
* Source/WebKit/UIProcess/FrameProcess.cpp:
(WebKit::FrameProcess::FrameProcess):
* Source/WebKit/UIProcess/FrameProcess.h:
(WebKit::FrameProcess::sharedProcessMainFrameSite const):
(WebKit::FrameProcess::create):
* Source/WebKit/UIProcess/WebBackForwardList.cpp:
(WebKit::WebBackForwardList::backForwardAddItemShared):
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::prepareForProvisionalLoadInProcess):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::launchProcess):
(WebKit::WebPageProxy::initializeWebPage):
(WebKit::WebPageProxy::receivedNavigationActionPolicyDecision): Fixed a subtle bug that this code was
using a wrong BrowsingContextGroup to find the shared process when the navigation is supposed to swap
the BrowsingContextGroup.
(WebKit::WebPageProxy::continueNavigationInNewProcess):
(WebKit::WebPageProxy::triggerBrowsingContextGroupSwitchForNavigation):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebProcessCache.cpp: Improved the logging code in this file to use
SENSITIVE_LOG_STRING of PRIVATE_LOG_STRING since it logs the site name.
(WebKit::WebProcessCache::canCacheProcess const):
(WebKit::WebProcessCache::addProcess):
(WebKit::WebProcessCache::evictAtRandomIfNeeded): Extracted from addProcess. We try to evict a shared
process before trying to evict the "main" process since the main process needs to launch sooner.
(WebKit::WebProcessCache::takeProcess):
(WebKit::WebProcessCache::takeSharedProcess): Added.
* Source/WebKit/UIProcess/WebProcessCache.h:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::processForSite): Added the logic to reuse a cached shared process.
(WebKit::WebProcessPool::createWebPage):
(WebKit::WebProcessPool::processForNavigation):
(WebKit::WebProcessPool::prepareProcessForNavigation):
(WebKit::WebProcessPool::processForNavigationInternal):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::addSharedProcessDomain): Added.
(WebKit::WebProcessProxy::didStartUsingProcessForSiteIsolation):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::siteIsolatedViewWithSharedProcess):
(TestWebKitAPI::(SiteIsolation, SharedProcessBasicWebProcessCache)): Added
(TestWebKitAPI::(SiteIsolation, SharedProcessWithResourceLoadStatistics)):
(TestWebKitAPI::(SiteIsolation, SharedProcessAfterClick)):
(TestWebKitAPI::(SiteIsolation, SharedProcessAfterKeyDown)):
(TestWebKitAPI::(SiteIsolation, SharedProcessAfterUserInteractionInSharedProcesss)):
(TestWebKitAPI::(SiteIsolation, SharedProcessWebProcessCacheSharedProcessForSiteWithUserInteraction)):
Added.

Canonical link: https://commits.webkit.org/300964@main
